services/cd-service/pkg/repository/repository.go

```diff
@@ -2330,6 +2336,321 @@ func (s *State) WriteAllCommitEvents(ctx context.Context, transaction *sql.Tx, d
+For really unusual configurations, these will be logged and not returned.
+*/
+func CalculateWarnings(_ context.Context, appName string, groups []*api.EnvironmentGroup) []*api.Warning {
+	result := make([]*api.Warning, 0)
+	for e := 0; e < len(groups); e++ {
+		group := groups[e]
+		for i := 0; i < len(groups[e].Environments); i++ {
+			env := group.Environments[i]
+			if env == nil || env.Config == nil || env.Config.Upstream == nil || env.Config.Upstream.Environment == nil {
+				// if the env has no upstream, there's nothing to warn about
+				continue
+			}
+			upstreamEnvName := env.Config.GetUpstream().Environment
+			upstreamEnv := getEnvironmentByName(groups, *upstreamEnvName)
+			if upstreamEnv == nil {
+				// this is already checked on startup and therefore shouldn't happen here
+				continue
+			}
+
+			appInEnv := env.Applications[appName]
+			if appInEnv == nil {
+				// appName is not deployed here, ignore it
+				continue
+			}
+			versionInEnv := appInEnv.Version
+			appInUpstreamEnv := upstreamEnv.Applications[appName]
+			if appInUpstreamEnv == nil {
+				// appName is not deployed upstream... that's unusual!
+				var warning = api.Warning{
+					WarningType: &api.Warning_UpstreamNotDeployed{
+						UpstreamNotDeployed: &api.UpstreamNotDeployed{
+							UpstreamEnvironment: *upstreamEnvName,
+							ThisVersion:         versionInEnv,
+							ThisEnvironment:     env.Name,
+						},
+					},
+				}
+				result = append(result, &warning)
+				continue
+			}
+			versionInUpstreamEnv := appInUpstreamEnv.Version
+
+			if versionInEnv > versionInUpstreamEnv && len(appInEnv.Locks) == 0 {
+				var warning = api.Warning{
+					WarningType: &api.Warning_UnusualDeploymentOrder{
+						UnusualDeploymentOrder: &api.UnusualDeploymentOrder{
+							UpstreamVersion:     versionInUpstreamEnv,
+							UpstreamEnvironment: *upstreamEnvName,
+							ThisVersion:         versionInEnv,
+							ThisEnvironment:     env.Name,
+						},
+					},
+				}
+				result = append(result, &warning)
+			}
+		}
+	}
+	return result
+}
+
+func deriveUndeploySummary(appName string, groups []*api.EnvironmentGroup) api.UndeploySummary {
+	var allNormal = true
+	var allUndeploy = true
+	for _, group := range groups {
+		for _, environment := range group.Environments {
+			var app, exists = environment.Applications[appName]
+			if !exists {
+				continue
+			}
+			if app.Version == 0 {
+				// if the app exists but nothing is deployed, we ignore this
+				continue
+			}
+			if app.UndeployVersion {
+				allNormal = false
+			} else {
+				allUndeploy = false
+			}
+		}
+	}
+	if allUndeploy {
+		return api.UndeploySummary_UNDEPLOY
+	}
+	if allNormal {
+		return api.UndeploySummary_NORMAL
+	}
+	return api.UndeploySummary_MIXED
+
+}
+
+func (s *State) UpdateOneAppEnv(ctx context.Context, transaction *sql.Tx, appName string, envName string, configParam *config.EnvironmentConfig) (*api.Environment_Application, error) {
```

Comment:
Same as above. The function name is confusing.

Reply:
done